futures-await-async-macro/src/lib.rs

```diff
@@ -224,42 +239,79 @@ where F: FnOnce(&Type) -> proc_macro2::TokenStream
 
 #[proc_macro_attribute]
 pub fn async(attribute: TokenStream, function: TokenStream) -> TokenStream {
-    // Handle arguments to the #[async] attribute, if any
     let (boxed, send) = match &attribute.to_string() as &str {
-        "( boxed )" => (true, false),
-        "( boxed_send )" => (true, true),
+        "( anchored )" => (true, false),
+        "( anchored_send )" => (true, true),
         "" => (false, false),
-        _ => panic!("the #[async] attribute currently only takes `boxed` as an arg"),
+        _ => panic!("the #[async] attribute currently only takes `anchored` as an arg"),
     };
 
-    async_inner(boxed, function, quote_cs! { ::futures::__rt::gen }, |output| {
+    async_inner(boxed, true, function, quote_cs! { ::futures::__rt::gen_pinned }, |output, lifetimes| {
         // TODO: can we lift the restriction that `futures` must be at the root of
         //       the crate?
         let output_span = first_last(&output);
         let return_ty = if boxed && !send {
             quote_cs! {
-                ::futures::__rt::std::boxed::Box<::futures::Future<
+                ::futures::__rt::PinBox<::futures::__rt::Future<
                     Item = <! as ::futures::__rt::IsResult>::Ok,
                     Error = <! as ::futures::__rt::IsResult>::Err,
                 >>
             }
         } else if boxed && send {
             quote_cs! {
-                ::futures::__rt::std::boxed::Box<::futures::Future<
+                ::futures::__rt::PinBox<::futures::__rt::Future<
                     Item = <! as ::futures::__rt::IsResult>::Ok,
                     Error = <! as ::futures::__rt::IsResult>::Err,
                 > + Send>
             }
+        } else {
+            quote_cs! {
+                impl ::futures::__rt::MyStableFuture<!> + #( #lifetimes + )*
+            }
+        };
+        let return_ty = respan(return_ty.into(), &output_span);
+        replace_bang(return_ty, &output)
+    })
+}
+
+#[proc_macro_attribute]
+pub fn async_move(attribute: TokenStream, function: TokenStream) -> TokenStream {
+    // Handle arguments to the #[async_move] attribute, if any
+    let (boxed, send) = match &attribute.to_string() as &str {
+        "( boxed )" => (true, false),
+        "( boxed_send )" => (true, true),
+        "" => (false, false),
+        _ => panic!("the #[async_move] attribute currently only takes `boxed` as an arg"),
+    };
+
+    async_inner(boxed, false, function, quote_cs! { ::futures::__rt::gen_move }, |output, lifetimes| {
+        // TODO: can we lift the restriction that `futures` must be at the root of
```

Comment:
I'm not familiar with proc-macros 2.0, but in macro-rules macros you could have a `#[doc(hidden)]` re-export of `futures-core` which you access using `$crate::futures::...`. Would something like that work here?

Reply:
Not my comment, but no, there's no `$crate::` in the current proc macros.